modules/weaver-tests/src/smithy4s/weavertests/WeaverTests.scala

```diff
@@ -0,0 +1,219 @@
+      testCase: HttpRequestTestCase,
+      inputFromDocument: Document.Decoder[I]
+  ) =
+    // todo: protocol check
+    test(endpoint.id.toString + "(server): " + testCase.id) {
+      Deferred[IO, I]
+        .flatMap { requestDeferred =>
+          val fakeImpl: smithy4s.Monadic[Alg, IO] =
+            service.transform(new smithy4s.Interpreter[Op, IO] {
+              def apply[I_, E_, O_, SE_, SO_](
+                  op: Op[I_, E_, O_, SE_, SO_]
+              ): IO[O_] = {
+                val (in, endpointInternal) = service.endpoint(op)
+
+                if (endpointInternal.id == endpoint.id)
+                  requestDeferred.complete(in.asInstanceOf[I]) *>
+                    IO.raiseError(new NotImplementedError)
+                else IO.raiseError(new Throwable("Wrong endpoint called"))
+              }
+            })
+
+          server(fakeImpl)
+            .liftTo[IO]
+            .flatMap(_.orNotFound.run(makeRequest(baseUri, testCase)))
+            .attemptNarrow[NotImplementedError] *>
+            requestDeferred.get.flatMap { foundInput =>
+              inputFromDocument
+                .decode(testCase.params.getOrElse(sys.error("no params: todo")))
+                .liftTo[IO]
+                .map { decodedInput =>
+                  // todo: derive Eq from schemas?
+                  implicit val eqq: Eq[I] = Eq.fromUniversalEquals
+                  assert.eql(foundInput, decodedInput)
+                }
+            }
+        }
+    }
+
+  private def makeClientTest[I, E, O, SE, SO](
+      endpoint: Endpoint[Op, I, E, O, SE, SO],
+      testCase: HttpRequestTestCase,
+      inputFromDocument: Document.Decoder[I]
+  ) = {
+    // todo: protocol check
+    test(endpoint.id.toString + "(client): " + testCase.id) {
+
+      Deferred[IO, Request[IO]]
+        .flatMap { requestDeferred =>
+          val theClient: IO[smithy4s.Monadic[Alg, IO]] = client(
+            {
+
+              HttpApp[IO] { req =>
+                // Save consumed stream for later reuse
+                req.body.compile.toVector
+                  .map(fs2.Stream.emits(_))
+                  .map(req.withBodyStream(_))
+                  .flatMap(requestDeferred.complete(_))
+                  .as(Response[IO]())
+              }
+            },
+            baseUri
+          ).liftTo[IO]
+
+          theClient
+            .flatMap { client =>
+              inputFromDocument
+                .decode(
+                  testCase.params.getOrElse(sys.error("no params, todo"))
+                )
+                .liftTo[IO]
+                .flatMap { in =>
+                  service
+                    .asTransformation(
+                      client
+                    )
+                    .apply(endpoint.wrap(in))
+                    // We don't expect the response to parse, as it's empty.
+                    .attemptNarrow[PayloadError]
+                }
+            } *> requestDeferred.get
```

Comment:
I think we can use WriterT instead

Reply:
nah I'm cool with Deferred, it's actually less confusing to most people